src/index.ts

```diff
@@ -283,37 +280,72 @@ export class Upload extends Pumpify {
   }
 
   private async startUploading() {
-    const reqOpts = {
-      method: 'PUT',
-      url: this.uri!,
-      headers: {
-        'Content-Range': 'bytes ' + this.offset + '-*/' + this.contentLength
-      }
-    };
+    // The buffer stream allows us to keep chunks in memory
+    // until we are sure we can successfully resume the upload.
+    const bufferStream = this.bufferStream || new PassThrough();
 
-    const bufferStream = this.bufferStream = new PassThrough();
+    // The offset stream allows us to analyze each incoming
+    // chunk to analyze it against what the upstream API already
+    // has stored for this upload.
     const offsetStream = this.offsetStream =
-        new PassThrough({transform: this.onChunk.bind(this)});
+        new Transform({transform: this.onChunk.bind(this)});
+
+    // The delay stream gives us a chance to catch the response
+    // from the API request before we signal to the user that
+    // the upload was successful.
     const delayStream = new PassThrough();
 
-    const requestStream = await this.getRequestStream(reqOpts);
-    this.setPipeline(bufferStream, offsetStream, requestStream, delayStream);
+    // The request library (authClient.request()) requires the
+    // stream to be sent within the request options.
+    const requestStreamEmbeddedStream = new PassThrough();
 
-    // wait for "complete" from request before letting the stream finish
     delayStream.on('prefinish', () => {
+      // Pause the stream from finishing so we can process the
+      // response from the API.
       this.cork();
     });
 
-    requestStream.on('complete', resp => {
-      if (resp.statusCode < 200 || resp.statusCode > 299) {
+    // Process the API response to look for errors that came in
+    // the response body.
+    this.on('response', (resp: GaxiosResponse) => {
+      if (resp.data.error) {
+        this.destroy(resp.data.error);
+        return;
+      }
+
+      if (resp.status < 200 || resp.status > 299) {
         this.destroy(new Error('Upload failed'));
         return;
       }
 
-      this.emit('metadata', resp.body);
+      this.emit('metadata', resp.data);
       this.deleteConfig();
+
+      // Allow the stream to continue naturally so the user's
+      // "finish" event fires.
       this.uncork();
     });
+
+    this.setPipeline(bufferStream, offsetStream, delayStream);
+
+    this.pipe(requestStreamEmbeddedStream);
+
+    this.once('restart', () => {
```

Comment:
@callmehiphop A new event is fired when `restart()` is called, which will let us unpipe the request stream.